database/rrdhost.c

```diff
@@ -1181,6 +1201,22 @@ int rrdhost_set_system_info_variable(struct rrdhost_system_info *system_info, ch
         freez(system_info->kernel_name);
         system_info->kernel_name = strdupz(value);
     }
+    else if(!strcmp(name, "NETDATA_SYSTEM_CPU_LOGICAL_CPU_COUNT")){
```

Comment:
I don't like how rrdhost_set_system_info_variable() always frees whatever is in every system_info field, relying on the undocumented fact that if something hasn't been set yet it should be NULL.

Reply:
Why is this a problem? What do you think will go wrong?

Reply:
Somebody could call `rrdhost_set_system_info_variable()` without having zeroed the structure.

Also, `freez()` is not documented to handle NULL arguments.

Reply:
One solution is to document both functions. E.g.

```
// system_info fields must be allocated or NULL
rrdhost_set_system_info_variable(struct rrdhost_system_info *system_info...)

// If ptr is NULL, no operation is performed.
freez()

```

Reply:
I understood your point of view @mfundul , and I will add an `if` before to free the information.
We did not change this before, because this was the original code and we only brought the necessary variables.

Reply:
Something analogous what I am bringing is also done inside health

Reply:
Wait, you don't need an `if`, what do you mean exactly?

Reply:
I commit now the changes, please, take a look.
This pattern that I brought here is used inside `health/health_config.c`, but it was not present in this file.

Reply:
I saw the change, this doesn't solve anything:
1. If the `system_info` structure is NOT zeroed out you will corrupt the memory anyway
2. If the `system_info` structure IS zeroed out then you gain nothing because `freez()` was already safe because it calls `free()` inside that is safe according to documentation.

Reply:
1 . System_info is zeroed inside `stream/rrdpush.c` and `daemon/main.c`, there is not way Netdata will arrive in this point with any value.
2 . Sure, but we have two different "patterns" on our code, one used in health, that is possible to be reloaded during runtime, and another here. I added the `if` for we have only one pattern in our code. I agree that it was not necessary to add `ifs` now, but it can be in the future.

Reply:
1. That's the point, we may add calls in the future.
2. The `if`s don't add anything, cause if it's NULL `freez()` works anyway. If it's random garbage, then the `freez()` will corrupt the memory equally with or without an `if`.

Reply:
I agreed with `1.` and `2.`, so I am keeping the `if` for case we start calling this function during runtime, and I also added the suggested documentation.

Reply:
Please remove the ifs again - they are redundant as there is a check inside `freez()`. @mfundul While your point is technically true it is not helpful in this case. The whole of the Netdata codebase operates on the assumption that `callocz` is used to allocate data-structures specifically so that they are zero'd out as their initial state. There is no code-path in which the contents of that structure are uninitialized. Whilst you may disagree with this policy as it exists globally within the entire code-base - this PR is not the place to address it.

Reply:
Are you sure ?

https://github.com/netdata/netdata/blob/4dc217a48a0b337e1f0baa13faa93a12c575a500/libnetdata/libnetdata.c#L171

:D

Reply:
Ah, geez... yes I had checked that and saw there is a check -- BUT failed to see that works only if NETDATA_LOG_ALLOCATIONS is set -- thanks @thiagoftsm :smile: